### PR TITLE
gh-93065: Fix HAMT to iterate correctly over 7-level deep trees

### PR DESCRIPTION
**Pull request by 1st1**

### CODE REVIEW

File: Include/internal/pycore_hamt.h [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
-#define _Py_HAMT_MAX_TREE_DEPTH 7
+#define _Py_HAMT_MAX_TREE_DEPTH 8
```
Comment (vstinner) [inline Include/internal/pycore_hamt.h:20]: It lacks an important comment like: "8 should be enough for everyone", famous last words.

**Comment by molaxx** [comment]
Hey @1st1 as I found and fixed this bug (https://github.com/MagicStack/immutables/pull/85) can I open this PR, I'll really like to have a merge in CPython on my name.

**Comment by 1st1** [comment]
@molaxx Eli, the commit is a bit more involved now as I added a comment explaining the fix + a test. I'd prefer to keep this PR and merge it.

That said, this change is yours, so I changed the author of the commit to your name and also added you to Misc/ACKS.

Now since you're the author, you need to accept our CLA here: https://github.com/python/cpython/pull/93066#issuecomment-1133827234

File: Include/internal/pycore_hamt.h (hunk 1 above)
Comment (1st1) [reply Include/internal/pycore_hamt.h:20]: good point actually, i added an explainer

**Comment by molaxx** [comment]
@1st1 Thanks!

File: Include/internal/pycore_hamt.h [hunk 2]
```diff
@@ -5,7 +5,19 @@
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
-#define _Py_HAMT_MAX_TREE_DEPTH 7
+
+/*
+HAMT tree is shaped by hashes of keys. Every group of 5 bits of a hash denotes
```
Comment (vstinner) [inline Include/internal/pycore_hamt.h:10]: You may mention "gh-93065" somewhere for details.

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for adding a comment explaining the 8 constant!

File: Include/internal/pycore_hamt.h (hunk 2 above)
Comment (ambv) [reply Include/internal/pycore_hamt.h:10]: Nah, the current comment describes the design exactly as intended. The fact that there was a bug before is kind of spurious.

File: Misc/NEWS.d/next/Core and Builtins/2022-05-21-23-21-37.gh-issue-93065.5I18WC.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Fix contextvars HAMT implementation to handle iteration over deep trees.
+
+The bug was discovered and fixed by Eli Libman. See [1] for more details.
+
+[1] https://github.com/MagicStack/immutables/issues/84
```
Comment (ambv) [inline Misc/NEWS.d/next/Core and Builtins/2022-05-21-23-21-37.gh-issue-93065.5I18WC.rst]: This doesn't look great in the actual change log later:
Screen Shot 2022-05-23 at 19 38 32

I'll change it to an inline link and land.

**Comment by ambv** [comment]
Backporting to all security branches as potential crashers are DoS.

File: Misc/NEWS.d/next/Core and Builtins/2022-05-21-23-21-37.gh-issue-93065.5I18WC.rst (hunk 3 above)
Comment (ambv) [reply Misc/NEWS.d/next/Core and Builtins/2022-05-21-23-21-37.gh-issue-93065.5I18WC.rst]: After:
Screen Shot 2022-05-23 at 20 19 33

**Review by ambv — APPROVED** [review]

**Comment by 1st1** [comment]
Thanks so much Lukasz!!

**Pull request merged**